### PR TITLE
support per-file sync

### PR DESCRIPTION
**Pull request by MichaelBrim**
### Description

This updates the `unifyfs_sync()` method in the client to support sync of either all outstanding writes or writes to a specific file. The server rpc handler did not require any modifications to accommodate this change.

The biggest change to how the client operates is that we no longer put writes in the shmem index as they happen. This avoids any interleaving of writes to different files. Instead, each sync rpc only handles the writes for a single file, and we always generate the contents of the shmem write index from each file's `extents_sync` segment tree at the time of the sync. This behavior matches what we were doing before when `UNIFYFS_FLATTEN_WRITES` was enabled (which was the default mode of operation).

Also included were fixes for several places where the wrong in/out structure types were used in the rpc invokers and handlers.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Testing (addition of new tests or update to current tests)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the UnifyFS code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted.

### CODE REVIEW

**Comment by adammoody** [comment]
Thanks, @MichaelBrim .  This will be a nice addition.

File: client/src/unifyfs-fixed.c [hunk 1]
```diff
@@ -270,84 +120,100 @@ void unifyfs_rewrite_index_from_seg_tree(void)
...
+        /* is this the target file? */
+        if ((target_fid != -1) && (fid != target_fid)) {
+            continue;
+        }
 
-    /* flushed, clear buffer and refresh number of entries
-     * and number remaining */
-    clear_index();
+        unifyfs_filemeta_t* meta = unifyfs_get_meta_from_fid(fid);
+        if ((NULL == meta) || (meta->fid != fid)) {
```
Comment (tonyhutter) [inline client/src/unifyfs-fixed.c:177]: `(meta->fid != fid)` <- is this check needed?

Comment (MichaelBrim) [reply client/src/unifyfs-fixed.c:177]: When fds/fids are re-used, the fid field is reset to -1. This is just a paranoid check to make sure the meta is for the current fid.

Comment (MichaelBrim) [reply client/src/unifyfs-fixed.c:177]: I wanted to put the check in the `unifyfs_get_meta_from_fid()` function itself, but it's used to do get the filemeta pointer when a fid is first allocated (before fid is set).

**Comment by MichaelBrim** [comment]
I just finished testing the various combination of configs on Summitdev. Everything seems to be working well (and write might even be faster). This is ready for review.

File: common/src/unifyfs_client_rpcs.h [hunk 2]
```diff
@@ -102,9 +116,9 @@ DECLARE_MARGO_RPC_HANDLER(unifyfs_metaget_rpc)
...
  *
- * given app_id and client_id as input, read all write extents
- * from client index in shared memory and insert corresponding
- * key/value pairs into our global metadata */
+ * given a client identified by (app_id, client_id) as input, read the write
+ * extents for one or more of the client's files from the shared memory index
+ * and update the global metadata for the file(s) */
 MERCURY_GEN_PROC(unifyfs_sync_in_t,
                  ((int32_t)(app_id))
                  ((int32_t)(client_id)))
```
Comment (sandrain) [inline common/src/unifyfs_client_rpcs.h:124]: @MichaelBrim because the sync rpc is per-file basis, can we add a gfid as an additional parameter here?

Comment (MichaelBrim) [reply common/src/unifyfs_client_rpcs.h:124]: We could, but the gfid can easily be retrieved from the shmem data, so I'm not sure what is the benefit.

Comment (sandrain) [reply common/src/unifyfs_client_rpcs.h:124]: We can possibly remove the repeated gfid from `unifyfs_index_t`, but maybe we can address later. We might need it back for different sync operations, if ever.

Comment (MichaelBrim) [reply common/src/unifyfs_client_rpcs.h:124]: I avoided changing the server rpc handler  (`rm_cmd_sync`) mostly to help avoid any conflicts with work going on in margotreenew.

File: client/src/unifyfs-fixed.c (hunk 1 above)
Comment (tonyhutter) [reply client/src/unifyfs-fixed.c:177]: Got it, thanks for the info

File: client/src/unifyfs-fixed.c [hunk 3]
```diff
@@ -129,126 +59,46 @@ static void clear_index(void)
...
+                 file_pos + length - 1,
+                 log_pos);
 
     return UNIFYFS_SUCCESS;
 }
 
 /*
  * Remove all entries in the current index and re-write it using the write
- * metadata stored in all the file's seg_trees.  This only re-writes the
+ * metadata stored in target files' seg_trees.  This only re-writes the
```
Comment (tonyhutter) [inline client/src/unifyfs-fixed.c:101]: "file's"

Comment (MichaelBrim) [reply client/src/unifyfs-fixed.c]: Thanks, I fixed the comment.

File: client/src/unifyfs-fixed.c [hunk 4]
```diff
@@ -50,76 +50,6 @@
...
-    off_t next_offset = next_idx->file_pos;
-    if (prev_offset != next_offset) {
-        /* index values are not contiguous, can't combine */
-        return UNIFYFS_SUCCESS;
-    }
-
-    /* got contiguous bytes in the same file,
-     * check whether both index values fall in the same slice */
-    off_t prev_slice = prev_idx->file_pos / slice_size;
-    off_t next_slice = next_idx->file_pos / slice_size;
```
[the remark this answers is not in the document]
Comment (MichaelBrim) [reply client/src/unifyfs-fixed.c:80]: latest push removes this

File: client/src/unifyfs-fixed.c [hunk 5]
```diff
@@ -270,84 +120,100 @@ void unifyfs_rewrite_index_from_seg_tree(void)
...
-     * if we're using that optimization */
-    if (unifyfs_flatten_writes) {
-        unifyfs_rewrite_index_from_seg_tree();
-    }
+    int ret = UNIFYFS_SUCCESS;
 
-    /* if there are no index entries, we've got nothing to sync */
-    if (*unifyfs_indices.ptr_num_entries == 0) {
-        return UNIFYFS_SUCCESS;
+    if (target_fid == -1) {
```
[the remark this answers is not in the document]
Comment (MichaelBrim) [reply client/src/unifyfs-fixed.c]: latest push figures out if any of the extents wrote to spillover, and does a logio sync if so

**Comment by MichaelBrim** [comment]
@adammoody I addressed your comments

**Pull request merged**